File: samples/Cpp/TestCpp/Classes/AccelerometerTest/AccelerometerTest.cpp
#include "AccelerometerTest.h"
#include "../testResource.h"
#include "cocos2d.h"

#define FIX_POS(_pos, _min, _max) \
    if (_pos < _min)        \
    _pos = _min;        \
else if (_pos > _max)   \
    _pos = _max;        \

//------------------------------------------------------------------
//
// AccelerometerTest
//
//------------------------------------------------------------------
AccelerometerTest::AccelerometerTest(void)
: _lastTime(0.0)
{
}

AccelerometerTest::~AccelerometerTest(void)
{
    _ball->release();
}

std::string AccelerometerTest::title()
{
    return "AccelerometerTest";
}

void AccelerometerTest::onEnter()
{
    Layer::onEnter();

    setAccelerometerEnabled(true);


    LabelTTF* label = LabelTTF::create(title().c_str(), "Arial", 32);
    addChild(label, 1);
    label->setPosition( ccp(VisibleRect::center().x, VisibleRect::top().y-50) );

    _ball = Sprite::create("Images/ball.png");
    _ball->setPosition(ccp(VisibleRect::center().x, VisibleRect::center().y));
    addChild(_ball);

    _ball->retain();
}

void AccelerometerTest::didAccelerate(Acceleration* pAccelerationValue)
{
//     double fNow = pAccelerationValue->timestamp;
// 
//     if (_lastTime > 0.0)
//     {
//         Point ptNow = convertToUI
//     }
// 
//     _lastTime = fNow;

    Director* pDir = Director::sharedDirector();

    /*FIXME: Testing on the Nexus S sometimes _ball is NULL */
    if ( _ball == NULL ) {
        return;
    }

    Size ballSize  = _ball->getContentSize();

    Point ptNow  = _ball->getPosition();
    Point ptTemp = pDir->convertToUI(ptNow);

    ptTemp.x += pAccelerationValue->x * 9.81f;
    ptTemp.y -= pAccelerationValue->y * 9.81f;

    Point ptNext = pDir->convertToGL(ptTemp);
    FIX_POS(ptNext.x, (VisibleRect::left().x+ballSize.width / 2.0), (VisibleRect::right().x - ballSize.width / 2.0));
    FIX_POS(ptNext.y, (VisibleRect::bottom().y+ballSize.height / 2.0), (VisibleRect::top().y - ballSize.height / 2.0));
    _ball->setPosition(ptNext);
}

//------------------------------------------------------------------
//
// AccelerometerTestScene
//
//------------------------------------------------------------------
void AccelerometerTestScene::runThisTest()
{
    Layer* pLayer = new AccelerometerTest();
    addChild(pLayer);
    pLayer->release();

    Director::sharedDirector()->replaceScene(this);
}

File: samples/Lua/TestLua/Resources/luaScript/AccelerometerTest/AccelerometerTest.lua
local function AccelerometerMainLayer()

    local function title()
      return "AccelerometerTest"
    end
    local layer = cc.Layer:create()

    layer:setAccelerometerEnabled(true)

    local label = cc.LabelTTF:create(title(), "Arial", 32)
    layer:addChild(label, 1)
    label:setPosition( cc.p(VisibleRect:center().x, VisibleRect:top().y - 50) )

    local ball = cc.Sprite:create("Images/ball.png")
    ball:setPosition(cc.p(VisibleRect:center().x, VisibleRect:center().y))
    layer:addChild(ball)

    ball:retain()
    
    local function didAccelerate(x,y,z,timestamp)

      if nil == ball then
        return
      end

      local director = cc.Director:getInstance()
      local szBall  = ball:getContentSize()
      local ptNowX,ptNowY = ball:getPosition()
      local ptTemp = director:convertToUI(cc.p(ptNowX,ptNowY)) 

      ptTemp.x  = ptTemp.x + x * 9.81 
      ptTemp.y  = ptTemp.y - y * 9.81

      local ptNext = director:convertToGL(cc.p(ptTemp.x,ptTemp.y))

      local minX  = math.floor(VisibleRect:left().x + szBall.width / 2.0)
      local maxX  = math.floor(VisibleRect:right().x - szBall.width / 2.0)
      if ptNext.x <   minX then
        ptNext.x = minX
      elseif ptNext.x > maxX then
        ptNext.x = maxX
      end
      
      local minY  = math.floor(VisibleRect:bottom().y + szBall.height / 2.0)
      local maxY  = math.floor(VisibleRect:top().y   - szBall.height / 2.0)
      if ptNext.y <   minY then
        ptNext.y = minY
      elseif ptNext.y > maxY then
        ptNext.y = maxY
      end
      
      ball:setPosition(cc.p(ptNext.x , ptNext.y))
    end
    
    layer:registerScriptAccelerateHandler(didAccelerate)   
    
    return layer
end


function AccelerometerMain()
	cclog("AccelerometerMain")
	local scene = cc.Scene:create()
	scene:addChild(AccelerometerMainLayer())
	scene:addChild(CreateBackMenuItem())
	return scene
end
